gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java

```diff
@@ -2082,6 +2110,68 @@ private boolean canIgnoreExceptionForEmptyObject(
     return false;
   }
 
+  @Override
+  public boolean isHnBucket(URI src) throws IOException {
+    String bucketName = src.getAuthority();
+    Boolean isEnabled = cache.getIfPresent(bucketName);
+    if (isEnabled != null) {
+      return isEnabled;
+    }
+
+    String prefix = src.getPath().substring(1);
+
+    StorageControlClient storageControlClient = lazyGetStorageControlClient();
+    GetStorageLayoutRequest request =
+        GetStorageLayoutRequest.newBuilder()
+            .setPrefix(prefix)
+            .setName(StorageLayoutName.format("_", bucketName))
+            .build();
+
+    try (ITraceOperation to = TraceOperation.addToExistingTrace("getStorageLayout.HN")) {
+      StorageLayout storageLayout = storageControlClient.getStorageLayout(request);
+      boolean result =
+          storageLayout.hasHierarchicalNamespace()
+              && storageLayout.getHierarchicalNamespace().getEnabled();
+
+      logger.atInfo().log("Checking if %s is HN enabled returned %s", src, result);
+
+      cache.put(bucketName, result);
+
+      return result;
+    }
+  }
+
+  private StorageControlClient lazyGetStorageControlClient() throws IOException {
+    if (this.storageControlClient == null) {
+      this.storageControlClient =
+          StorageControlClient.create(
+              StorageControlSettings.newBuilder()
+                  .setCredentialsProvider(FixedCredentialsProvider.create(this.credential))
+                  .build());
+    }
+
+    return this.storageControlClient;
+  }
+
+  @Override
+  public void renameHnFolder(URI src, URI dst) throws IOException {
```

Comment:
should we validate that both `src` and `dst` have same Authority i.e. bucket?

Reply:
Good point. I will fix this in a later change.